### PR TITLE
Aggregations Refactor: Refactor Moving Average Aggregation

### PR DESCRIPTION
**Pull request by colings86**

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/MovAvgPipelineAggregator.java [hunk 1]
```diff
@@ -277,32 +280,151 @@ public void doWriteTo(StreamOutput out) throws IOException {
...
+        /**
+         * Gets the GapPolicy to use on the output of this aggregation.
+         */
+        public GapPolicy gapPolicy() {
+            return gapPolicy;
+        }
+
+        protected ValueFormatter formatter() {
+            if (format != null) {
+                return ValueFormat.Patternable.Number.format(format).formatter();
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/MovAvgPipelineAggregator.java]: I don't know how often this is called, depending on this maybe it makes sense to store the formatter somewhere for later reuse unless `format` changes? Is only called a few times maybe not worth the trouble.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java [hunk 2]
```diff
@@ -107,11 +114,42 @@ public String getName() {
...
+        builder.field(MovAvgParser.MODEL.getPreferredName(), NAME_FIELD.getPreferredName());
+        builder.startObject(MovAvgParser.SETTINGS.getPreferredName());
+        builder.field("alpha", alpha);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public MovAvgModel readFrom(StreamInput in) throws IOException {
+        return new EwmaModel(in.readDouble());
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java]: Could you also use the Prototype here?

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java [hunk 3]
```diff
@@ -107,11 +114,42 @@ public String getName() {
...
+    }
+
+    @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeString(STREAM.getName());
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java]: Curious about where the stream name gets read on the receiving side. Maybe read/write could be changed to be more symetric, but I haven't completely checked the deserialization code.

**Comment by cbuescher** [comment]
Left a few small comments, would improve readability if readFrom/writeTo for the Models would be somehow symetric, but maybe that's hard to achieve with the current serialization infra. Otherwise looks good to me.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java]: I took another look at MovAvgModelStreams, and although I'm not completely sure it look a lot like what NamedWritable is doing, so I was wondering if Stream could be replaced by it.

**Comment by jpountz** [comment]
LGTM

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java (hunk 3 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/models/EwmaModel.java]: I opened https://github.com/elastic/elasticsearch/issues/15279 for this

**Pull request merged**